enhancements/ingress/set-delete-http-headers.md

```diff
@@ -0,0 +1,1031 @@
+the `haproxy.config` sections corresponding to the following backends: plain http backend, backend with TLS terminated at the edge, secure backend with re-encryption ([code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609)).
+
+The router sets this variable based on the Route's `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type` or `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type` field values.
+`ServiceAliasConfig` is populated by [the router's `createServiceAliasConfig` method](https://github.com/openshift/router/blob/3065f6583f3925328fbdbfe95e3bc7bb7a084d33/pkg/router/template/router.go#L908-L935)
+
+```go
+// ServiceAliasConfig is a route for a service.  Uniquely identified by host + path.
+type ServiceAliasConfig struct {
+    ...
+    // For route based setting of http response headers
+    HTTPResponseHeaders []HTTPHeader
+
+    // For route based setting of http request headers
+    HTTPRequestHeaders []HTTPHeader
+}
+```
+```go
+// HTTPHeader specifies an HTTP header that should be set/replace/append
+type HTTPHeader struct {
+// Name specifies an HTTP header name.
+Name string
+
+// Value specifies the header value.
+Value string
+
+// Action specifies the action to be performed.
+Action string
+}
+```
+
+In this case the ordering of actions will be maintained as we will be having access to the RouteSpec inside the router. So, we will run a loop
+on either `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type` or `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type` which will have the desired
+sequence. 
+
+If request and response headers are provided in the API then first we run loop through request headers and then response headers.
+
+C) library-go code will be having validations for checking Route API fields for OCP.
+   We will be creating an immutable/hard-coded deny-list of HTTP response headers in [library-go](https://github.com/openshift/library-go/blob/2a52b7ce52f99e2ed185718380b43734877f2439/pkg/route/validation/validation.go#L48-L141)
+   which won't be allowed to be set or deleted. 
+
+   For the initial enhancement, the HSTS `strict-transport-security` header name is included in this deny-list.
+   
+   Only 64 items in request and response header list will be allowed.
+   
+   Minimum length of the header value would have to be greater than 0 and less than 4096.
+   
+   Sample fetchers for response dynamic header values allowed are "res.hdr" and "ssl_c_der".
+   
+   Sample fetchers for request dynamic header values allowed are "req.hdr" and "ssl_c_der".
+   
+   Converters allowed for request and response are "lower" and "base64".
+   
+  Actions like Set/Delete for headers can't be set for passthrough routes.
```

Comment:
```suggestion
  Actions like Set/Delete for headers are ignored for passthrough routes.
```

Reply:
I believe "can't be set" is more accurate—validation catches attempts to specify custom headers for passthrough routes:
```go
    // spec is the desired state of the route
    // +kubebuilder:validation:XValidation:rule="!has(self.tls) || self.tls.termination != 'passthrough' || !has(self.httpHeaders)",message="header actions are not permitted when tls termination is passthrough."
    Spec RouteSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
```
If you want this to be more explicit, how about this?
```suggestion
  Validation prohibits configuring any header actions on passthrough routes.
```

Reply:
Does `||` mean something other than `or` here?

Reply:
it is a rule. condition or

Reply:
i still think `Actions like Set/Delete for headers can't be set for passthrough routes.` is more better

Reply:
Validation prohibits configuring any header actions on passthrough routes.  added this as looks more better